File: src/backend/app/services/ocp/cluster_topology_heal.py
"""Heal cluster topology drift (legacy migration ghosts, wrong member parentage)."""

from __future__ import annotations

import copy
import re

LEGACY_GHOST_CLUSTER_ID = "ocp"
LEGACY_GHOST_NODE_ID = "cluster-ocp"

_DEPLOY_ONLY_CLUSTER_FIELDS = frozenset(
    {
        "_generatedInstallConfig",
        "_generatedAgentConfig",
        "_generatedPullThroughItms",
        "controlPlaneDisks",
        "workerDisks",
    }
)

_MEMBER_ID_RE = re.compile(r"^(.+)-(cp|worker)-\d+$")


def _is_legacy_migration_ghost(cluster: dict) -> bool:
    return (
        cluster.get("id") == LEGACY_GHOST_CLUSTER_ID
        and cluster.get("nodeId") == LEGACY_GHOST_NODE_ID
        and cluster.get("baseDomain") in (None, "", "ocp.local")
    )


def _strip_deploy_only(cluster: dict) -> dict:
    out = {k: v for k, v in cluster.items() if k not in _DEPLOY_ONLY_CLUSTER_FIELDS}
    return out


def _infer_cluster_id_from_member_id(vm_id: str) -> str | None:
    m = _MEMBER_ID_RE.match(vm_id or "")
    return m.group(1) if m else None


def _clusters_are_legacy_ghost_only(clusters: list) -> bool:
    return bool(clusters) and all(_is_legacy_migration_ghost(c) for c in clusters)


def freeze_deployed_cluster_ocp_versions(current: dict, deployed: dict) -> None:
    """Revert ocpVersion on clusters already in deployed_topology.

    OCP version is fixed at install; canvas edits must not survive reconfigure.
    """
    dep_versions = {
        c["id"]: c.get("ocpVersion") or ""
        for c in (deployed.get("clusters") or [])
        if c.get("id")
    }
    if not dep_versions:
        return
    for cluster in current.get("clusters") or []:
        cid = cluster.get("id")
        if cid not in dep_versions:
            continue
        frozen = dep_versions[cid]
        if frozen:
            cluster["ocpVersion"] = frozen


def _reconcile_canvas_clusters(
    canvas_clusters: list,
    deployed_clusters: list,
    nodes: list,
) -> list:
    stripped_deployed = [_strip_deploy_only(c) for c in deployed_clusters]
    has_ghost = any(_is_legacy_migration_ghost(c) for c in canvas_clusters)
    base = [c for c in canvas_clusters if not _is_legacy_migration_ghost(c)]

    if (has_ghost or not base) and stripped_deployed:
        seen = {c.get("id") for c in base}
        for dc in stripped_deployed:
            if dc.get("id") not in seen:
                base.append(dc)
                seen.add(dc.get("id"))

    by_node = {c.get("nodeId"): c for c in base if c.get("nodeId")}
    for node in nodes:
        if node.get("type") != "clusterNode":
            continue
        if node.get("id") == LEGACY_GHOST_NODE_ID:
            continue
        data = node.get("data") or {}
        cid = data.get("clusterId")
        if not cid or node.get("id") in by_node or cid in {c.get("id") for c in base}:
            continue
        dep = next(
            (
                c
                for c in stripped_deployed
                if c.get("id") == cid or c.get("nodeId") == node.get("id")
            ),
            None,
        )
        entry = {
            "id": cid,
            "name": data.get("name") or (dep or {}).get("name") or cid,
            "nodeId": node.get("id"),
            "type": data.get("type") or (dep or {}).get("type") or "sno",
            "controlPlane": data.get("controlPlane")
            if data.get("controlPlane") is not None
            else (dep or {}).get("controlPlane", 1),
            "workers": data.get("workers")
            if data.get("workers") is not None
            else (dep or {}).get("workers", 0),
            "baseDomain": data.get("baseDomain")
            or (dep or {}).get("baseDomain")
            or "local",
            "apiVip": data.get("apiVip") or (dep or {}).get("apiVip"),
            "ingressVip": data.get("ingressVip") or (dep or {}).get("ingressVip"),
            "ocpVersion": (dep or {}).get("ocpVersion") or "",
            "networkIds": (dep or {}).get("networkIds") or data.get("networkIds") or [],
        }
        base.append(entry)

    for node in nodes:
        if node.get("type") != "clusterNode":
            continue
        data = node.get("data") or {}
        cid = data.get("clusterId")
        if not cid:
            continue
        cluster = next((c for c in base if c.get("id") == cid), None)
        if not cluster:
            continue
        if data.get("controlPlane") is not None:
            cluster["controlPlane"] = data["controlPlane"]
        if data.get("workers") is not None:
            cluster["workers"] = data["workers"]

    return [c for c in base if not _is_legacy_migration_ghost(c)]


def _resolve_member_cluster_id(
    node: dict, clusters: list, deployed_ids: set[str]
) -> str | None:
    data = node.get("data") or {}
    by_id = {c.get("id"): c for c in clusters if c.get("id")}
    cid = data.get("clusterId")
    inferred = _infer_cluster_id_from_member_id(node.get("id", ""))

    if inferred and inferred in by_id:
        return inferred

    if cid == LEGACY_GHOST_CLUSTER_ID and any(
        not _is_legacy_migration_ghost(c) for c in clusters
    ):
        name = str(data.get("name") or "")
        if re.match(r"^cp-\d+$", name):
            sno = [
                c
                for c in clusters
                if c.get("type") == "sno" and not _is_legacy_migration_ghost(c)
            ]
            deployed_sno = [c for c in sno if c.get("id") in deployed_ids]
            if len(deployed_sno) == 1:
                return deployed_sno[0].get("id")
            if len(sno) == 1:
                return sno[0].get("id")
        if inferred:
            return inferred

    if cid in by_id:
        return cid
    return inferred if inferred in by_id else None


def _heal_membership(nodes: list, clusters: list, deployed_clusters: list) -> None:
    by_id = {c.get("id"): c for c in clusters if c.get("id")}
    node_by_id = {n.get("id"): n for n in nodes}
    deployed_ids = {str(c.get("id")) for c in deployed_clusters if c.get("id")}

    for node in nodes:
        ntype = node.get("type")
        data = node.get("data") or {}

        if ntype == "vmNode" and (data.get("os") == "rhcos" or data.get("clusterId")):
            cid = _resolve_member_cluster_id(node, clusters, deployed_ids)
            cluster = by_id.get(cid) if cid else None
            if not cluster:
                continue
            node["parentId"] = cluster.get("nodeId")
            data["clusterId"] = cluster.get("id")
            continue

        if (
            ntype == "storageNode"
            and node.get("parentId") == LEGACY_GHOST_NODE_ID
            and any(not _is_legacy_migration_ghost(c) for c in clusters)
        ):
            owner_id = re.sub(r"-disk-\d+$", "", node.get("id", ""))
            owner = node_by_id.get(owner_id)
            if not owner:
                continue
            cid = _resolve_member_cluster_id(owner, clusters, deployed_ids)
            cluster = by_id.get(cid) if cid else None
            if cluster:
                node["parentId"] = cluster.get("nodeId")


def _order_parents_before_children(nodes: list) -> list:
    id_to_index = {n.get("id"): i for i, n in enumerate(nodes)}
    out = list(nodes)
    for node in nodes:
        parent_id = node.get("parentId")
        if not parent_id:
            continue
        child_id = node.get("id")
        pi = id_to_index.get(parent_id)
        ci = id_to_index.get(child_id)
        if pi is None or ci is None or ci > pi:
            continue
        child = out.pop(ci)
        # Rebuild index map after pop
        id_to_index = {n.get("id"): i for i, n in enumerate(out)}
        pi = id_to_index.get(parent_id)
        if pi is None:
            out.append(child)
        else:
            out.insert(pi + 1, child)
        id_to_index = {n.get("id"): i for i, n in enumerate(out)}
    return out


def heal_cluster_topology(
    topology: dict,
    *,
    deployed_clusters: list | None = None,
) -> dict:
    """Return a copy with clusters[], membership, and node order repaired."""
    topo = copy.deepcopy(topology or {})
    nodes = topo.get("nodes") or []
    deployed = deployed_clusters or []

    clusters = _reconcile_canvas_clusters(
        topo.get("clusters") or [],
        deployed,
        nodes,
    )
    if clusters:
        topo["clusters"] = clusters

    if clusters and any(not _is_legacy_migration_ghost(c) for c in clusters):
        nodes = [n for n in nodes if n.get("id") != LEGACY_GHOST_NODE_ID]

    _heal_membership(nodes, clusters, deployed)
    topo["nodes"] = _order_parents_before_children(nodes)
    return topo


def seed_topology_clusters_from_deployed(
    topology: dict, deployed_clusters: list
) -> dict:
    """Seed/heal canvas clusters from deployed_topology (no legacy ghost)."""
    return heal_cluster_topology(topology, deployed_clusters=deployed_clusters)

File: src/frontend/src/components/canvas/clusterTopologyHeal.ts
import type { Node, Edge } from "@xyflow/react";
import type { ClusterConfig } from "@/stores/canvasStore";
import { orderChildAfterParent } from "@/components/canvas/clusterMembership";
import { healClusterBoundaryWidths } from "@/components/canvas/clusterMaterialize";

/** Legacy lazy-migration artifact — must not coexist with modern multi-cluster boxes. */
export const LEGACY_GHOST_CLUSTER_ID = "ocp";
export const LEGACY_GHOST_NODE_ID = "cluster-ocp";

const DEPLOY_ONLY_CLUSTER_FIELDS = [
  "_generatedInstallConfig",
  "_generatedAgentConfig",
  "controlPlaneDisks",
  "workerDisks",
] as const;

export function isLegacyMigrationGhost(c: ClusterConfig): boolean {
  return (
    c.id === LEGACY_GHOST_CLUSTER_ID &&
    c.nodeId === LEGACY_GHOST_NODE_ID &&
    (c.baseDomain === "ocp.local" || !c.baseDomain)
  );
}

function stripDeployOnlyClusterFields(dc: ClusterConfig): ClusterConfig {
  const c = { ...dc } as Record<string, unknown>;
  for (const f of DEPLOY_ONLY_CLUSTER_FIELDS) delete c[f];
  return c as unknown as ClusterConfig;
}

/** Infer cluster id from a generated member VM id (`<clusterId>-cp-0`, etc.). */
export function inferClusterIdFromMemberId(vmId: string): string | null {
  const m = vmId.match(/^(.+)-(cp|worker)-\d+$/);
  return m ? m[1] : null;
}

function clusterConfigFromBoundaryNode(
  node: Node,
  deployed?: ClusterConfig,
): ClusterConfig {
  const d = node.data as Record<string, unknown>;
  const id =
    (d.clusterId as string) ||
    deployed?.id ||
    node.id.replace(/^cluster-/, "");
  return {
    id,
    name: (d.name as string) || deployed?.name || id,
    nodeId: node.id,
    type: (d.type as string) || deployed?.type || "sno",
    controlPlane: (d.controlPlane as number) ?? deployed?.controlPlane ?? 1,
    workers: (d.workers as number) ?? deployed?.workers ?? 0,
    controlPlaneCpu: deployed?.controlPlaneCpu ?? 8,
    controlPlaneMemory: deployed?.controlPlaneMemory ?? 16384,
    controlPlaneDisk: deployed?.controlPlaneDisk ?? 120,
    workerCpu: deployed?.workerCpu ?? 4,
    workerMemory: deployed?.workerMemory ?? 8192,
    workerDisk: deployed?.workerDisk ?? 100,
    baseDomain: (d.baseDomain as string) || deployed?.baseDomain || "local",
    apiVip: (d.apiVip as string) || deployed?.apiVip || "",
    ingressVip: (d.ingressVip as string) || deployed?.ingressVip || "",
    ocpVersion: deployed?.ocpVersion || "",
    pullThroughRegistry: deployed?.pullThroughRegistry,
    usePullThroughRegistry: deployed?.usePullThroughRegistry,
    networkIds: deployed?.networkIds || (d.networkIds as string[]) || [],
    recert: deployed?.recert,
    monitorHealth: deployed?.monitorHealth,
    configureBastionBrowser: deployed?.configureBastionBrowser,
    installOnDeploy: deployed?.installOnDeploy,
  };
}

/**
 * Rebuild `clusters[]` from deployed_topology, boundary nodes, and the canvas
 * list — dropping the legacy migration ghost when real clusters exist.
 */
export function reconcileCanvasClusters(
  canvasClusters: ClusterConfig[],
  deployedClusters: ClusterConfig[] | undefined,
  nodes: Node[],
): ClusterConfig[] {
  const strippedDeployed = (deployedClusters || []).map(stripDeployOnlyClusterFields);
  const hasGhost = canvasClusters.some(isLegacyMigrationGhost);
  const hasRealCanvas = canvasClusters.some((c) => !isLegacyMigrationGhost(c));

  // The legacy ghost (cluster-ocp) is a synthetic migration artifact only when a
  // REAL cluster exists to replace it (another canvas cluster or a deployed
  // cluster) — that is the case it was created to heal (it "stole" members from
  // real boxes). When "ocp"/cluster-ocp is the SOLE cluster it is a legitimate
  // single-cluster OCP project (every fresh sno/compact/standard template makes
  // exactly this id), not a ghost — dropping it deletes the real cluster box and
  // its member VM. Only drop the ghost when there is something to replace it.
  const hasReplacement = hasRealCanvas || strippedDeployed.length > 0;
  const dropGhost = hasGhost && hasReplacement;

  let base: ClusterConfig[] = dropGhost
    ? canvasClusters.filter((c) => !isLegacyMigrationGhost(c))
    : [...canvasClusters];

  if ((hasGhost || base.length === 0) && strippedDeployed.length > 0) {
    for (const dc of strippedDeployed) {
      if (!base.some((c) => c.id === dc.id)) {
        base.push(dc);
      }
    }
  }

  const realBoundaryNodes = nodes.filter(
    (n) =>
      n.type === "clusterNode" &&
      n.id !== LEGACY_GHOST_NODE_ID &&
      (n.data as Record<string, unknown>)?.clusterId,
  );

  for (const boundary of realBoundaryNodes) {
    const d = boundary.data as Record<string, unknown>;
    const cid = d.clusterId as string;
    if (base.some((c) => c.id === cid || c.nodeId === boundary.id)) continue;
    const dep = strippedDeployed.find(
      (c) => c.id === cid || c.nodeId === boundary.id,
    );
    base.push(clusterConfigFromBoundaryNode(boundary, dep));
  }

  if (!hasRealCanvas && base.length === 0 && !hasGhost) {
    return canvasClusters;
  }

  const filtered = dropGhost
    ? base.filter((c) => !isLegacyMigrationGhost(c))
    : base;
  for (const boundary of realBoundaryNodes) {
    const d = boundary.data as Record<string, unknown>;
    const cid = d.clusterId as string;
    const cluster = filtered.find((c) => c.id === cid);
    if (!cluster) continue;
    if (d.controlPlane !== undefined && d.controlPlane !== null) {
      cluster.controlPlane = d.controlPlane as number;
    }
    if (d.workers !== undefined && d.workers !== null) {
      cluster.workers = d.workers as number;
    }
  }
  return filtered;
}

function resolveMemberClusterId(
  node: Node,
  clusters: ClusterConfig[],
  deployedIds: Set<string>,
): string | null {
  const d = node.data as Record<string, unknown>;
  const byId = new Map(clusters.map((c) => [c.id, c]));
  const cid = d.clusterId as string | undefined;

  const inferred = inferClusterIdFromMemberId(node.id);
  if (inferred && byId.has(inferred)) {
    return inferred;
  }

  if (
    cid === LEGACY_GHOST_CLUSTER_ID &&
    clusters.some((c) => !isLegacyMigrationGhost(c))
  ) {
    const name = String(d.name || "");
    if (/^cp-\d+$/.test(name)) {
      const sno = clusters.filter(
        (c) => c.type === "sno" && !isLegacyMigrationGhost(c),
      );
      const deployedSno = sno.filter((c) => deployedIds.has(c.id));
      if (deployedSno.length === 1) return deployedSno[0].id;
      if (sno.length === 1) return sno[0].id;
    }
    if (inferred) return inferred;
  }

  if (cid && byId.has(cid)) return cid;
  return inferred && byId.has(inferred) ? inferred : null;
}

/** Point RHCOS members (and their hidden disks) at the right cluster boundary. */
export function healClusterMembership(
  nodes: Node[],
  clusters: ClusterConfig[],
  deployedClusters?: ClusterConfig[],
): Node[] {
  const byId = new Map(clusters.map((c) => [c.id, c]));
  const deployedIds = new Set(
    (deployedClusters || []).map((c) => String(c.id || "")).filter(Boolean),
  );

  return nodes.map((n) => {
    const d = n.data as Record<string, unknown>;

    if (n.type === "vmNode" && (d.os === "rhcos" || d.clusterId)) {
      const cid = resolveMemberClusterId(n, clusters, deployedIds);
      const cluster = cid ? byId.get(cid) : undefined;
      if (!cluster) return n;
      return {
        ...n,
        parentId: cluster.nodeId,
        extent: "parent" as const,
        draggable: false,
        data: { ...d, clusterId: cluster.id },
      };
    }

    if (
      n.type === "storageNode" &&
      n.parentId === LEGACY_GHOST_NODE_ID &&
      clusters.some((c) => !isLegacyMigrationGhost(c))
    ) {
      const diskOwner = n.id.replace(/-disk-\d+$/, "");
      const owner = nodes.find((x) => x.id === diskOwner);
      if (!owner) return n;
      const cid = resolveMemberClusterId(owner, clusters, deployedIds);
      const cluster = cid ? byId.get(cid) : undefined;
      if (!cluster) return n;
      return { ...n, parentId: cluster.nodeId };
    }

    return n;
  });
}

export function dedupeNodeIds(nodes: Node[]): Node[] {
  const seen = new Set<string>();
  return nodes.filter((n) => {
    if (seen.has(n.id)) return false;
    seen.add(n.id);
    return true;
  });
}

/** React Flow requires each parent boundary to appear before its children. */
export function orderClusterParentsBeforeChildren(nodes: Node[]): Node[] {
  let ordered = [...nodes];
  for (const n of nodes) {
    if (!n.parentId) continue;
    ordered = orderChildAfterParent(ordered, n.id, n.parentId);
  }
  return ordered;
}

function removeLegacyGhostBoundary(
  nodes: Node[],
  clusters: ClusterConfig[],
): Node[] {
  if (!clusters.some((c) => !isLegacyMigrationGhost(c))) return nodes;
  return nodes.filter((n) => n.id !== LEGACY_GHOST_NODE_ID);
}

export interface HealClusterTopologyInput {
  nodes: Node[];
  edges: Edge[];
  clusters: ClusterConfig[];
  deployedClusters?: ClusterConfig[];
}

export interface HealClusterTopologyResult {
  nodes: Node[];
  edges: Edge[];
  clusters: ClusterConfig[];
}

/**
 * Repair cluster metadata drift: legacy migration ghosts, orphaned boundary
 * boxes, wrong member parentId/clusterId, duplicate ids, and parent ordering.
 */
export function healClusterTopology(
  input: HealClusterTopologyInput,
): HealClusterTopologyResult {
  const clusters = reconcileCanvasClusters(
    input.clusters,
    input.deployedClusters,
    input.nodes,
  );
  let nodes = dedupeNodeIds(input.nodes);
  nodes = removeLegacyGhostBoundary(nodes, clusters);
  nodes = healClusterMembership(nodes, clusters, input.deployedClusters);
  nodes = healClusterBoundaryWidths(nodes);
  nodes = orderClusterParentsBeforeChildren(nodes);
  return { nodes, edges: input.edges, clusters };
}
